### ISSUE TITLE
Add possibility to bypass dependencies 

### ISSUE THREAD

**Issue by David-Lannion** [mentalisttraceur/robotframework-dependencylibrary#6]
As a tester, I want to be able to run/rerun specific Test alone even if it has dependencies

Idea :
Add possibility to not take into account the case :
`if status is None:
        raise AssertionError(message + ' not found.')`

**Comment by David-Lannion** [comment]
I wanted to make a proposition but I can't make a branch to push it

**Comment by mentalisttraceur** [comment]
Oops sorry accidentally clicked "close" button while drafting a comment.

**Comment by David-Lannion** [comment]
Here is my use case :
My full run is something like :

virtual_car.robot

> *** Test Cases ***
> Check ressources to build
>     [...]
> 
> Make motor
>     [...]
> 
> Make car structure
>     [...]
> 
> Put car together
>     [...]

If my run Fail or then developping, some part are already done.
For example, I got all parts of the car, I just want to "put it together", i will only run the specific Test with
`robot -t Put_car_together virtual_car.robot`

If I use dependencies, it will not work
However when the full run is running, if some part are skipped or failed, I don't want the Test to run (so dependencies are usefull here)

**Comment by David-Lannion** [comment]
Not sure is this example is making my use case clear

**Comment by mentalisttraceur** [comment]
Yep, that clarified it for me, thank you!

**Comment by David-Lannion** [comment]
The way I see it's implementation usage is the following :

> *** Settings ***
> Resource      DependencyLibrary  none_as_pass=True

**Comment by David-Lannion** [comment]
In code :

```
def  _depends_on(status_map, dependency_type, name, none_as_pass=False):
     message = 'Dependency not met: ' + dependency_type + ' ' + repr(name)
     status = status_map.get(name.lower(), None)
    if status is None:
        if not none_as_pass:
            raise AssertionError(message + ' not found.')
        else:
            return "Not Found but pass"
    if status is Ellipsis:
        raise AssertionError(message + ' mid-execution.')
    if status == 'PASS':
        return 'PASS'
    if status == 'SKIP':
        raise _SkipExecution(message + ' was skipped.')
    assert status == 'FAIL', message + ' has status ' + repr(status) + '.'
    raise AssertionError(message + ' failed.')
 
 
 class DependencyLibrary(object):
     ROBOT_LISTENER_API_VERSION = 3
     ROBOT_LIBRARY_SCOPE = 'GLOBAL'
 
     def __init__(self, none_as_pass=False):
         self.ROBOT_LIBRARY_LISTENER = self
         self._test_status_map = {}
         self._suite_status_map = {}
         self.none_as_pass = none_as_pass
    [...]
      def depends_on_test(self, name):
           return _depends_on(self._test_status_map, 'test case', name, self.none_as_pass)
```
Note : I added returns to have infos in log.html


usage exemple :
`robot --variable SKIP_DEPENDENCIES:True ./test.robot`
```
*** Settings ***
Library           DependencyLibrary    none_as_pass=${SKIP_DEPENDENCIES}

*** Test Cases ***
Passing Test
    No operation

Skipped Test
    Skip    This test is skipped for some reason.

Dependent Test
    Depends on test    Passing Test
    Depends on test    PLOP    # not existing Test, seen as status=None
```

**Comment by David-Lannion** [comment]
I edited last comment with the things I made and an exemple

**Comment by mentalisttraceur** [comment]
I suggest using a wrapper keyword like this:

```robot
*** Keywords ***
Depends on test
    [Arguments]  ${test_name}
    ${test_name_repr} =  Evaluate  repr($test_name)
    TRY
        DependencyLibrary.depends on test  ${test_name}
    EXCEPT  Dependency not met: test case ${test_name_repr} not found.
        No operation
```

I'll think more about adding an in-library feature like you suggest. My instinct is that a wrapper keyword like this is the righter way but I'm happy to hear arguments or other suggestions.

In the meantime let me know if the above keyword solves your problem for the time being!

**Comment by mentalisttraceur** [comment]
I haven't tested that keyword yet, just sketched out the idea as soon as I found the time to finish it.

**Comment by David-Lannion** [comment]
I implemented the changes presented on my "exemple" comment and used it that way

**Comment by mentalisttraceur** [comment]
I tentatively decided that this use-case/convenience is important/useful enough to support in the library, I'm just still thinking about what the best way is.

**Comment by mentalisttraceur** [comment]
Ideally it would be automatic - you wouldn't have to do anything and the library would just "know" when you're re-running a subset of tests after a previous run of existing tests.

Right now I'm thinking: if there's a way to detect that robot is being invoked on only a subset of the tests, and that the named tests exist but aren't included in the run, then that seems like a great user experience.

This would take some looking into, and I'd need to think carefully before committing to this if Robot doesn't provide that information, because then it would require re-reading the test files inside this library.

A simpler version is if we can just detect that robot is being invoked with specific tests specified, period. No trying to figure out what other tests are in there, just assume that if the invocation specifies specific tests, unknown tests are actually okay to treat as passing.

Going in the other direction, a more "perfect" version of this knows which tests have ran and passed in previous runs. This would require reading data saved between runs (and possibly writing that data if Robot doesn't write it for us), including possibly multiple runs, to know which depended-on tests have succeeded previously.

That seems... worth scouting a little, but feels like it's probably more complexity than justified. I'm imagining other various ways of doing this too. In a simpler world I would just say "look if a test is leaving behind results which you can use on a later run, just use those results instead of depending on the test", but that loses the "skip" convenience.

tempted to just change the behavior - no flag, just make it a warning instead of an error.

I'll have to test it later. if I go that way I want it to be a warning that's very visible by default.

Because it's the usual idempotency vs error problem where both behaviors are possibly useful.

(This would be a breaking change so of course that release would come with a major version bump, so that people actually relying on that error behavior and taking advantage of SemVer don't have problems.)

**Comment by mentalisttraceur** [comment]
Sorry. Accidentally pressed the "close" button again. I hate touch screens which don't require some pressure before committing to an input. It's way too easy to have a moment of fumbling and do something with the lightest unintended touch that would never trigger a physical mechanism.

**Comment by mentalisttraceur** [comment]
Ideally it would be automatic - you wouldn't have to do anything and the library would just "know" when you're re-running a subset of tests after a previous run of existing tests.

Right now I'm thinking: if there's a way to detect that robot is being invoked on only a subset of the tests, and that the named tests exist but aren't included in the run, then that seems like a great user experience.

This would take some looking into, and I'd need to think carefully before committing to this if Robot doesn't provide that information, because then it would require re-reading the test files inside this library.

A simpler version is if we can just detect that robot is being invoked with specific tests specified, period. No trying to figure out what other tests are in there, just assume that if the invocation specifies specific tests, unknown tests are actually okay to treat as passing.

Going in the other direction, a more "perfect" version of this knows which tests have ran and passed in previous runs. This would require reading data saved between runs (and possibly writing that data if Robot doesn't write it for us), including possibly multiple runs, to know which depended-on tests have succeeded previously.

That seems... worth scouting a little, but feels like it's probably more complexity than justified. It's also still fragile to stuff like test renaming between runs.

I'm also imagining other various ways of doing this too. In a simpler world I would just say "look, if a test is leaving behind results which you can use on a later run, just test for things being in that state instead of depending on the test", but that loses the "skip" convenience - one nice thing dependency library does is let skips cascade.

Another thought I was having was to just change the behavior - no flag, just make it a warning instead of an error, always. But if I go that way I want it to be a warning that's very visible by default. (This would be a breaking change so of course that release would come with a major version bump, so that people actually relying on that error behavior and taking advantage of SemVer don't have problems.) I have to carefully think about this because

1. a typo'ed dependency or a dependency on a test that has been renamed is an outright mistake in the test - the test is *wrong* and ideally we don't "fail open" by running wrong code for users;
2. a dependency on a test that was deleted is dead code that needs to be cleaned up, but arguably valid, coherent, and sensible - ideally we help and incentivize cleanup but force it; and
3. a dependency on a test that ran in a previous partial run, during a partial continuation re-run, is clearly correct and useful - ideally we make that work smoothly.

Basically this is yet another instance of the almost universal idempotency vs error problem - both interpretations of "this doesn't exist" are useful. A warning is good enough for the 2nd and 3rd case. But maybe a clear visible warning is good enough for the 1st case as well?

And then of course there's the idea we've already touched on: a flag for the library. The nice thing about this is that you have a decent middle-ground: you can specify it at file scope, which avoids some repetition at per-keyword level, while having the simplicity of letting the user communicate to the library when it's appropriate.

A command-line argument or environment variable (instead of or as an alternative to a library parameter) would kinda provide all the same benefits with the trade-off of more esse if you want to regularly flip it per-invocation but more repetitiveness if you're manually invoking it.

Then there's the possibility of adding a separate keyword, or a flag for the existing keyword, kinda like the wrapper keyword described so far. In some ways this seems nice because the "error or idempotent success?" dilemma doesn't have a universally good answer, and I can easily imagine wanting per-keyword granularity (because personally, opting out of errors is something I strongly prefer to do as narrowly as possible).

**Comment by mentalisttraceur** [comment]
A more critical look at:

> and I can easily imagine wanting per-keyword granularity

This is actually pretty rare though.

If I'm in a test and I have two `Depends on ...` keywords, when would I want one to tolerate a not-found test, but not the other?

Well, if one of those dependency tests is towards the beginning of a run and another is just before my test, and failures that force a re-run always are between them, I might reasonably think "oh I want to tolerate the absence of the first dependency but not the second".

But that's a different use-case isn't it? That's embracing *long term* lingering of dependency-not-found situations.

That's very different than on-the-fly iterative development/debugging/fixing of a test suite.

**Comment by David-Lannion** [comment]
> A more critical look at:
> 
> > and I can easily imagine wanting per-keyword granularity
> 
> This is actually pretty rare though.
> 
> If I'm in a test and I have two `Depends on ...` keywords, when would I want one to tolerate a not-found test, but not the other?
> 
> Well, if one of those dependency tests is towards the beginning of a run and another is just before my test, and failures that force a re-run always are between them, I might reasonably think "oh I want to tolerate the absence of the first dependency but not the second".
> 
> But that's a different use-case isn't it? That's embracing _long term_ lingering of dependency-not-found situations.
> 
> That's very different than on-the-fly iterative development/debugging/fixing of a test suite.

Then adding a keyword : "Depends on if exist" might solve the problem ?

**Comment by mentalisttraceur** [comment]
@David-Lannion right I was describing the use-cases where such a keyword would be good and also noticing that those are different from - and maybe less worth supporting than - the use-case you initially brought up.

**Comment by mentalisttraceur** [comment]
So I was thinking about just making the "test not found" case a warning in all situations.

If I do that, then these are our two protections against typos, stale test references, and so on:

The warning in `log.html`:
![Screenshot_20220610-005045](https://user-images.githubusercontent.com/6531708/172979123-ae0ed7e6-23d0-43bb-ac33-0d4783de21c3.png)

The CLI output from `robot`:
```
==============================================================================
Test                                                    
==============================================================================
Test
[ WARN ] Dependency not met: test case 'foo' not found.
Test
Test                                                                  | PASS |
------------------------------------------------------------------------------
Test                                                                  | PASS |
1 test, 1 passed, 0 failed
==============================================================================
```

**Comment by mentalisttraceur** [comment]
Today I almost made the decision to just go ahead and make it always a warning rather than a failure.

I think I already wrote the arguments for this above, but to recap:

1. We could look at "this test case does not exist" as more like an unused variable warning and less like a using undeclared variable error. (I still don't know if this is the best way to view it, but it's definitely a valid way.)
2. Ultimately the goal is to help developers catch mistakes.

**Comment by mentalisttraceur** [comment]
But one thing I gotta finish thinking through is that #5 might effect this:
I have a vision for an elegant composition of test-level parallelism through something like `pabot` with `DependencyLibrary`: you run with test level parallelism, and so long as you've used dependency library to properly annotate ALL your dependencies, it'll just work.

Of course to make that happen, there needs to be some feature work in `pabot` that I haven't gotten around to contributing, and then some more feature work in `DependencyLibrary`.

And on the `DependencyLibrary` side, that would effect how we handle test-does-not-exist situations.

**Comment by mentalisttraceur** [comment]
So what I picture in the beautiful `DependencyLibrary` and `pabot` synergy world, is something like this:

1. `pabot` starts launching every test it can shove into an executor.
2. `pabot` executes a test with `Depends on test  foo`.
3. If `foo` is already running, `DependencyLibrary` hits the "test in progress" case, and uses some `pabot` API to tell `pabot` to suspend this test case until that test is finished.
4. If `foo` is not found, `DependencyLibrary` hits the "test not found" case, and here's where we hit the challenge:
    * If test-not-found is an error, then DependencyLibrary can ask `pabot` to wait, just as in the test-in-progress case, and if `pabot` runs out of tests and the resume condition never becomes true, that's also an error, so that's consistent.
    * If test-not-found is a warning, then I guess we can ask `pabot` to wait until it sees it, but now if it runs all tests and still hasn't seen it, it needs to resume our test instead of failing.

**Comment by mentalisttraceur** [comment]
@David-Lannion what do you think about just having the "dependency not found" case just *always* be a warning? (Just a warning instead of causing a test failure?)

**Comment by mentalisttraceur** [comment]
I'm like 99% decided on moving forward with turning not-found dependencies into warnings. Already made the diff months ago, now I dug it up and committed it. Might release as soon as tomorrow.

If people speak up one way or the other great - it's technically a breaking change to the API (if anyone was relying on the not-found->test-fails behavior) so it'll release as v4.0.0.

If anyone really liked the old behavior they can keep using v3.* and I'll be open to hearing proposals for other ways of doing it.

I'm still fairly receptive to some of the other ideas brought up so far - global `robot` CLI option, environment variable, or if possible having the behavior predicated on if the depended-on test case is found in the suite but marked as "not run".

**Comment by mentalisttraceur** [comment]
Fixed in v4.0.0 - missing dependencies now just warn instead of failing the test.
